core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala

```diff
@@ -0,0 +1,135 @@
+  test("SPARK-30310: Test uncaught RuntimeException, exitOnUncaughtException = true") {
+    val process = getThrowerProcess(RuntimeExceptionThrower, exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.UNCAUGHT_EXCEPTION)
+  }
+
+  test("SPARK-30310: Test uncaught RuntimeException, exitOnUncaughtException = false") {
+    val process = getThrowerProcess(RuntimeExceptionThrower, exitOnUncaughtException = false)
+    assert(process.waitFor == 0)
+  }
+
+  test("SPARK-30310: Test uncaught OutOfMemoryError, exitOnUncaughtException = true") {
+    val process = getThrowerProcess(OutOfMemoryErrorThrower, exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+  test("SPARK-30310: Test uncaught OutOfMemoryError, exitOnUncaughtException = false") {
+    val process = getThrowerProcess(OutOfMemoryErrorThrower, exitOnUncaughtException = false)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException, exitOnUncaughtException = true") {
+    val process = getThrowerProcess(SparkFatalExceptionThrower, exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.UNCAUGHT_EXCEPTION)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException, exitOnUncaughtException = false") {
+    val process = getThrowerProcess(SparkFatalExceptionThrower, exitOnUncaughtException = false)
+    assert(process.waitFor == 0)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException (OOM), exitOnUncaughtException = true") {
+    val process = getThrowerProcess(SparkFatalExceptionWithOOMThrower,
+      exitOnUncaughtException = true)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+  test("SPARK-30310: Test uncaught SparkFatalException (OOM), exitOnUncaughtException = false") {
+    val process = getThrowerProcess(SparkFatalExceptionWithOOMThrower,
+      exitOnUncaughtException = false)
+    assert(process.waitFor == SparkExitCode.OOM)
+  }
+
+}
+
+// a thread that uses SparkUncaughtExceptionHandler, then throws the throwable
+class ThrowableThrowerThread(t: Throwable,
+    exitOnUncaughtException: Boolean) extends Thread {
+  override def run() {
+    Thread.setDefaultUncaughtExceptionHandler(
+      new SparkUncaughtExceptionHandler(exitOnUncaughtException))
+    throw t
+  }
+}
+
+// Objects to be invoked by spark-class for different Throwable types
+// that SparkUncaughtExceptionHandler handles.  spark-class will exit with
+// exit code dictated by either SparkUncaughtExceptionHandler (SparkExitCode)
+// or main() (0)
+
+object RuntimeExceptionThrower {
+  def main(args: Array[String]): Unit = {
+    val t = new ThrowableThrowerThread(new RuntimeException, if (args(0) == "true") true else false)
+    t.start()
+    t.join()
+    System.exit(0)
+  }
+}
+
+object OutOfMemoryErrorThrower {
+  def main(args: Array[String]): Unit = {
+    val t = new ThrowableThrowerThread(new OutOfMemoryError, if (args(0) == "true") true else false)
```

Comment:
what if args is empty?

Reply:
thanks for the suggestion!  modded the throwers to check args.length and exit with -1 if it's empty now.